DOCSP-30506 Async and Sync Runtimes (#54)
mongodb/docs-rust: source/fundamentals/async-sync.txt

      .. code-block:: toml

         [dependencies.mongodb]
         version = "2.6.0"
         default-features = false
         features = ["tokio-runtime"]

      Also install the crate by running:

      .. code-block:: none

         cargo add tokio

   .. tab:: async-std
      :tabid: asynchronous-api-async-std

      To use ``async-std``, add the following to your ``Cargo.toml``:

      .. code-block:: toml

         [dependencies.mongodb]
         version = "2.6.0"
         default-features = false
         features = ["async-std"]

      Also install the crate by running:

      .. code-block:: none

         cargo add async-std

For more information on installing both the driver and runtimes, see :ref:`rust-quick-start-download-and-install`.

Asynchronous Code Example (Tokio)
---------------------------------

Use the `Client() <{+api+}/struct.Client.html>`__ type for asynchronous programming.

The following code uses the ``task`` module from the ``tokio`` crate
to create separate, concurrent tasks for multiple data operations:

.. literalinclude:: /includes/fundamentals/code-snippets/async-sync-tokio.rs
   :language: rust

>    :dedent:

Author: I decided to use a new file since these two examples could diverge in the future.

.. _rust-async-sync-configure-sync:

Configure the Synchronous Runtime
---------------------------------

The driver also provides a blocking, synchronous API.
To enable this, add the ``sync`` or ``tokio-sync`` feature to your ``Cargo.toml``.

.. code-block:: toml

   [dependencies.mongodb]
   version = "2.6.0"
   default-features = false
   features = ["sync"]

Synchronous Code Example
------------------------

When using the synchronous runtime, use types from the ``mongodb::sync`` module to perform operations.
The synchronous API calls through to the asynchronous API internally,
so you can use ``sync`` types as you might use types from the asynchronous API.
The following code uses the ``sync`` module to insert data into a collection using the synchronous API.

> It uses the synchronous client type imported from the ``sync`` module on the first line.

Reviewer: S: combine these two lines for concision

.. _rust-async-sync:

=====================================
Asynchronous and Synchronous Runtimes
=====================================

.. contents:: On this page
   :local:

>    :backlinks: nones

Reviewer: ```suggestion
   :backlinks: none
```

   :depth: 2
   :class: singlecol

Overview
--------

> In this guide, you can learn how to configure and use the driver's asynchronous and synchronous APIs.

Reviewer: ```suggestion
In this guide, you can learn about the {+driver-short+}'s asynchronous and synchronous runtimes.
This guide explains how to enable the available runtimes and structure your code to use each API.
```

{+driver-short+} supports both ``tokio`` and ``async-std``, the two most popular asynchronous runtime crates.
By default, the driver uses ``tokio``, but you can explicitly choose a runtime
by specifying one of ``tokio-runtime`` or ``async-std-runtime`` feature flags in your ``Cargo.toml``.

> The driver also includes a synchronous API for simplicity, or when parallelism is not necessary.

Reviewer: S: replace this section with a list of the runtime that the driver supports. 

For example:
```
The driver supports the following asynchronous runtime crates:
- ``tokio`` (the default runtime)
- ``async-std-runtime``

To learn more about these runtimes, see the :ref:`...` section.

The driver supports the following synchronous runtime crates:
....

To learn more about these runtimes, see the :ref:`...` section.
```

Author: I did :) I added links to the subsections like you recommended. I personally think it's a bit nicer to have this overview flow instead of having the unordered list. But I'm happy to make the change if you think it'd be better.

Configure the Asynchronous Runtime
----------------------------------

> The driver uses ``tokio`` by default.

Reviewer: ```suggestion
The driver uses the ``tokio`` runtime by default.
```

> Use the `Client() <{+api+}/struct.Client.html>`__ type for asynchronous programming.

Reviewer: S: make a subsection for "Asynchronous Example" (same for sync section)

The following code uses the ``task`` module from the ``tokio`` crate to create separate, concurrent
tasks for multiple data operations:

.. literalinclude:: /includes/fundamentals/code-snippets/async-sync-tokio.rs
   :language: rust

>    :dedent:

Reviewer: S: This section could benefit from a short explanation of why/when users should use an async programming framework. [This page](https://rust-lang.github.io/async-book/01_getting_started/02_why_async.html) has some interesting notes. Also, it would be great to make some distinctions b/t the two async (and sync) crates. What makes them different? Can follow up with eng if needed.

Author: This gets into some pretty deep questions about application design that we might not want to cover in the drivers docs.

To use ``async-std``, add the following to your ``Cargo.toml``:

.. code-block:: toml

   [dependencies.mongodb]
   version = "2.6.0"
   default-features = false

>    features = ["async-std-runtime"]

Reviewer: S: I think the instructions on how to install the different crates should be first in this section. Maybe you can have some tabs on how to install each one.

> Enable the sync API
> ---------------------

Reviewer: I: this heading is mismatched from the other heading, and contains the wrong amount of underscores
```suggestion
Configure the Synchronous Runtime
------------------------------------
```

> The driver also provides a blocking, synchronous API.

Reviewer: S: same comment applies here. Why use sync? what makes the two sync runtimes different?

Author: In the overview, it says "for simplicity, or when parallelism is not necessary". Does that make sense?

> To enable this, add the ``sync`` or ``tokio-sync`` feature to your ``Cargo.toml``.

Reviewer: Show config for both runtimes

.. code-block:: toml

   [dependencies.mongodb]
   version = "2.6.0"
   default-features = false
   features = ["sync"]

> The sync-specific types can be imported from ``mongodb::sync`` (e.g. ``mongodb::sync::Client``).

Reviewer: S: Move into a new section

> The sync-specific types can be imported from ``mongodb::sync`` (e.g. ``mongodb::sync::Client``).

Reviewer: S: use active voice
```suggestion
When using the synchronous runtime, you must use types from the ``mongodb::sync`` module to perform operations. The synchronous API calls through to the asynchronous API internally, so you can use ``sync`` types as you might use types from the asynchronous API.
```

The following code uses the ``sync`` module to perform actions with the blocking API.

> Each request must finish before the next can begin.

Reviewer: ```suggestion
The driver waits for each request to complete before continuing.
```

.. literalinclude:: /includes/fundamentals/code-snippets/async-sync-sync.rs
   :language: rust
   :dedent:

Additional Information
----------------------

For more information about the concepts in this guide, see the following pages:

- :ref:`Performance Considerations <rust-performance>`
- `Asynchronous Programming in Rust <https://rust-lang.github.io/async-book/>`__
  
API Documentation

> -----------------

Reviewer: ```suggestion
~~~~~~~~~~~~~~~~~
```
